lib/WWW/Mechanize.pm

```diff
@@ -1454,6 +1433,59 @@ sub form_with_fields {
     }
 }
 
+
+=head2 $mech->form_with( $attr1 => $value1, $attr2 => $value2, ... )
+
+Searches for forms with arbitrary attribute/value pairs within the E<lt>formE<gt>
+tag.
+(Currently does not work for attribute C<action> due to implementation details
+of L<HTML::Form>.)
+When given more than one pair, all criteria must match.
+Using C<undef> as value means that the attribute in question may not be present.
+
+If it is found, the form is returned as an L<HTML::Form> object and set internally
+for later used with Mech's form methods such as C<L</field()>> and C<L</click()>>.
+
+Returns undef if no form is found.
+
+
+=cut
+
+sub form_with {
+    my ( $self, %spec ) = @_;
+
+    my @forms = $self->forms or return;
+    while ( my ( $attr, $expected_value ) = each %spec ) {
+        @forms = grep {
+            my $actual_value = $_->attr($attr);
+            if ( defined $expected_value ) {
+                defined $actual_value && $actual_value eq $expected_value;
+            }
+            else {
+                !defined $actual_value;
+            }
+        } @forms or return;
```

Comment:
That code simplifies down to:

        @forms = grep {
            !defined $expected_value || $_->attr($attr) eq $expected_value
        } @forms or return;